### ISSUE TITLE
[RFE] Statistical analysis of etcd logs using NumPy

### ISSUE THREAD

**Issue by arnegogala** [peterducai/openshift-etcd-suite#3]
A colleague recently shared this small Python code which is using NumPy to do some statistical analysis of all the "took too long" messages in etcd log. Could be a good idea to have this included in "openshift-etcd-suite" but will need adaption to more recent etcd log formats...

Example output:
~~~
$ analyze-etcd-logs-311.py etcd-XYZ.com/etcd/etcd/logs/current.log 
# Statistics
Count: 7511
Min: 100.014255
Max: 11397.865489000002
Average: 1030.7382432813208
Median: 197.599531
Std Dev: 1871.0100838908627
~~~

Code:
~~~
#!/usr/bin/env python3

import os.path
import sys
import re

import numpy as np
import matplotlib.pyplot as plt


etcd_logfile = sys.argv[1]
if not os.path.isfile(etcd_logfile):
    print('File does not exist')

times_list = []
with open(etcd_logfile, 'r') as logfile:
    for line in logfile:
        if "took too long" in line:

            # Get time from line
            time_string = re.search('\(.+\)', line).group()

            # Get rid of parentheses
            time_string = time_string[1:-1]

            if(re.match('[0-9]*\.[0-9]*ms', time_string)):
                time = time_string[0:-2]

            elif(re.match('[0-9]*\.[0-9]*s', time_string)):
                # Time is in seconds, convert it to ms
                time = time_string[0:-1]
                time = 1000 * float(time)

            times_list.append(time)

#for item in times_list:
#    print(item)

a = np.asarray(times_list).astype(np.float64)

print("# Statistics")
print("Count: {0}".format(len(times_list)))
print("Min: {0}".format(np.min(a)))
print("Max: {0}".format(np.max(a)))
print("Average: {0}".format(np.average(a)))
print("Median: {0}".format(np.median(a)))
print("Std Dev: {0}".format(np.std(a)))
~~~

**Comment by simonkrenger** [comment]
As the colleague in question, I am happy to also provide the updated script for OCP 4 (where the etcd log is in JSON format):

~~~
#!/usr/bin/env python3

import os.path
import sys
import re
import json

import numpy as np
import matplotlib.pyplot as plt


etcd_logfile = sys.argv[1]
if not os.path.isfile(etcd_logfile):
    print('File does not exist')

lines_skipped = 0
times_list = []
with open(etcd_logfile, 'r') as logfile:
    for line in logfile:
        if "request took too long" in line:

            # Remove timestamp at beginning of line
            # 2021-08-26T06:20:50.210590693Z {"level":"warn","ts":"20...
            line = line[31:]

            # Print lines for https://bugzilla.redhat.com/show_bug.cgi?id=2017004
            # etcd log sometimes contains invalid JSON
            #print(line)

            # Parse line
            try:
                line_json = json.loads(line)
            except:
                lines_skipped = lines_skipped + 1
                continue
            time_took = line_json["took"]

            # Process "ms" and "s"
            if(re.match('[0-9]*\.[0-9]*ms', time_took)):
                time = time_took[0:-2]

            elif(re.match('[0-9]*\.[0-9]*s', time_took)):
                # Time is in seconds, convert it to ms
                time = 1000 * float(time_took[0:-1])

            times_list.append(time)

#for item in times_list:
#    print(item)

a = np.asarray(times_list).astype(np.float64)

if lines_skipped > 0:
    print("")
    print("WARNING: LINES SKIPPED: {0}".format(lines_skipped))
    print("")

print("# Statistics")
print("Count: {0}".format(len(times_list)))
print("Min: {0}".format(np.min(a)))
print("Max: {0}".format(np.max(a)))
print("Average: {0}".format(np.average(a)))
print("Median: {0}".format(np.median(a)))
print("Std Dev: {0}".format(np.std(a)))
~~~

**Comment by peterducai** [comment]
Regarding etcd.sh I'm thinking to script everything in bash, simply to avoid any external tools. But I like idea of summary like

# Statistics
Count: 7511
Min: 100.014255
Max: 11397.865489000002
Average: 1030.7382432813208
Median: 197.599531
Std Dev: 1871.0100838908627

and this could be actually integrated container (which is now used by fio_testsuite). I will create additional issues and implement your ideas @simonkrenger @arnegogala into openshift-etcd-suite container.

**Comment by peterducai** [comment]
implemented in last commit, OCP 3.11 script works also on OCP 4 must-gather so i add it to etcd.sh error detection of 'took too long'.

**Comment by peterducai** [comment]
etcd_tooktoolong311.py used in etcd.sh
